---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 7,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>
  
## GD15.5:
```{r}
library(Seurat)
library(Signac)
library(ggplot2)
library(GenomicRanges)
set.seed(1234)


load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.rda")
colnames(rnor6)[1] <- "gene_id"

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5_res0.8.rda")
data <- RenameIdents(data, `0` = "Other cells", `1` = "Other cells",
                     `2` = "Other cells", `3` = "Other cells", `4` = "Other cells",
                     `5` = "Other cells", `6` = "Natural killer cells",
                     `7` = "Macrophage cells", `8` = "Macrophage cells",
                     `9` = "Natural killer cells", `10` = "Endothelial cells",
                     `11` = "Macrophage cells", `12` = "Other cells",
                     `13` = "Endothelial cells", `14` = "Smooth muscle cells",
                     `15` = "Endothelial cells", `16` = "Other cells",
                     `17` = "Smooth muscle cells", `18` = "Other cells",
                     `19` = "Natural killer cells", `20` = "Other cells",
                     `21` = "Other cells", `22` = "Other cells",
                     `23` = "Invasive trophoblast cells", `24` = "Other cells",
                     `25` = "Other cells", `26` = "Smooth muscle cells",
                     `27` = "Other cells", `28` = "Other cells")
Idents(data) <- factor(Idents(data), levels = c("Other cells", "Macrophage cells", "Smooth muscle cells",
                                                "Endothelial cells", "Natural killer cells", "Invasive trophoblast cells"))
rats.RNA <- data

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/2_mergedSamples/2_MACS2Peaks/GD15.5/15.5_1-2-3-merged.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")

rats <- unintegrated
DefaultAssay(rats) <- "combined"
mcols(rnor6.ranges, level="within")[, "gene_id"] <- mcols(rnor6.ranges, level="within")[, "ensembl_gene_id"]
Annotation(rats) <- rnor6.ranges
gene.activities <- GeneActivity(rats)
rats[['RNA']] <- CreateAssayObject(counts = gene.activities[, colnames(rats)])
rats <- NormalizeData(object = rats, assay = 'RNA', normalization.method = 'LogNormalize', scale.factor = median(rats$nCount_RNA))
DefaultAssay(rats) <- 'RNA'

transfer.anchors <- FindTransferAnchors(reference = rats.RNA, query = rats, reduction = 'cca')
predicted.labels <- TransferData(anchorset = transfer.anchors, refdata = Idents(rats.RNA), weight.reduction = rats[['lsi']], dims=2:20)

rats <- AddMetaData(object = rats, metadata= predicted.labels)

plot1 <- DimPlot(object = rats.RNA, cols = c("grey70", "orange3", "#718748", "#183D61", "#79704C", "red3")) + ggtitle('GD15.5 scRNA-seq')
plot2 <- DimPlot(object=rats, group.by = 'predicted.id', label = T, repel = T) + ggtitle('GD15.5 scATAC-seq')

plot1 + plot2

#save(rats, file="/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/gd15.5-ATACwithRNAlables.rda")

table(rats$predicted.id)
```

Correlation check:
```{r}
library(Seurat)
library(Signac)
library(Seurat)
library(S4Vectors)
library(GenomicRanges)
library(patchwork)
library(biomaRt)
library(ggplot2)
library(ggpubr)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
avgATAC <- AverageExpression(rats, assays = "RNA")
avgATAC <- as.data.frame(avgATAC)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5_res0.8.rda")
avgRNA <- AverageExpression(data, assays = "RNA")
avgRNA <- as.data.frame(avgRNA)
```

Trophoblast:
```{r}
avgRNA1 <- avgRNA$RNA.23
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Invasive.trophoblast.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-15.5-tbcCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Invasive Trophoblast Cells at gd 15.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```

Endothelial:
```{r}
avgRNA1 <- rowMeans(avgRNA[, c("RNA.10", "RNA.13", "RNA.15")])
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Endothelial.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(log10(avgRNA1+10^(-5)), log10(avgATAC1+10^(-5)), method = "spearman", exact = FALSE)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-15.5-endoCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=FALSE) + ggtitle("Correlation in Endothelial Cells at gd 15.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```
Macrophage:
```{r}
avgRNA1 <- rowMeans(avgRNA[, c("RNA.7", "RNA.8", "RNA.11")])
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Macrophage.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-15.5-macroCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Macrophage Cells at gd 15.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```
NK:
```{r}
avgRNA1 <- rowMeans(avgRNA[, c("RNA.6", "RNA.9", "RNA.19")])
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Natural.killer.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-15.5-nkCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Natural Killer Cells at gd 15.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```
Smooth muscle:
```{r}
avgRNA1 <- rowMeans(avgRNA[, c("RNA.14", "RNA.17", "RNA.26")])
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Smooth.muscle.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-15.5-smCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Smooth Muscle Cells at gd 15.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```

## GD19.5:
```{r}
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-4-5-6-7_res0.8.rda")
data <- data2
data <- RenameIdents(data, `0` = "Other cells", `1` = "Macrophage cells",
                     `2` = "Other cells", `3` = "Other cells", `4` = "Other cells",
                     `5` = "Other cells", `6` = "Invasive trophoblast cells",
                     `7` = "Macrophage cells", `8` = "Smooth muscle cells",
                     `9` = "Natural killer cells", `10` = "Endothelial cells",
                     `11` = "Macrophage cells", `12` = "Other cells",
                     `13` = "Other cells", `14` = "Macrophage cells",
                     `15` = "Other cells", `16` = "Other cells",
                     `17` = "Macrophage cells", `18` = "Macrophage cells",
                     `19` = "Other cells", `20` = "Endothelial cells",
                     `21` = "Other cells", `22` = "Other cells",
                     `23` = "Other cells", `24` = "Other cells",
                     `25` = "Other cells", `26` = "Smooth muscle cells")
Idents(data) <- factor(Idents(data), levels = c("Other cells", "Macrophage cells", "Smooth muscle cells",
                                                "Endothelial cells", "Natural killer cells", "Invasive trophoblast cells"))
rats.RNA <- data

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/2_mergedSamples/2_MACS2Peaks/GD19.5/19.5_4-5-6-merged.rda")
rats <- unintegrated
DefaultAssay(rats) <- "combined"
mcols(rnor6.ranges, level="within")[, "gene_id"] <- mcols(rnor6.ranges, level="within")[, "ensembl_gene_id"]
Annotation(rats) <- rnor6.ranges
gene.activities <- GeneActivity(rats)
rats[['RNA']] <- CreateAssayObject(counts = gene.activities[, colnames(rats)])
rats <- NormalizeData(object = rats, assay = 'RNA', normalization.method = 'LogNormalize', scale.factor = median(rats$nCount_RNA))
DefaultAssay(rats) <- 'RNA'

transfer.anchors <- FindTransferAnchors(reference = rats.RNA, query = rats, reduction = 'cca')
predicted.labels <- TransferData(anchorset = transfer.anchors, refdata = Idents(rats.RNA), weight.reduction = rats[['lsi']], dims=2:10)

rats <- AddMetaData(object = rats, metadata = predicted.labels)

plot1 <- DimPlot(object = rats.RNA, cols = c("grey70", "orange3", "#718748", "#183D61", "#79704C", "red3")) + ggtitle('GD19.5 scRNA-seq')
plot2 <- DimPlot(object=rats, group.by = 'predicted.id', label = T, repel = T) + ggtitle('GD19.5 scATAC-seq')

plot1 + plot2

#save(rats, file="/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/gd19.5-ATACwithRNAlables.rda")

table(rats$predicted.id)
```

Correlation check:
```{r}
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
avgATAC <- AverageExpression(rats, assays = "RNA")
avgATAC <- as.data.frame(avgATAC)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-4-5-6-7_res0.8.rda")
avgRNA <- AverageExpression(data2, assays = "RNA")
avgRNA <- as.data.frame(avgRNA)
```

Trophoblast:
```{r}
avgRNA1 <- avgRNA$RNA.6
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Invasive.trophoblast.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-19.5-tbcCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho") + ggtitle("Correlation in Invasive Trophoblast Cells at gd 19.5")
#dev.off()
```

Endothelial:
```{r}
avgRNA1 <- rowMeans(avgRNA[, c("RNA.10", "RNA.20")])
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Endothelial.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-19.5-endoCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Endothelial Cells at gd 19.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```
Macrophage:
```{r}
avgRNA1 <- rowMeans(avgRNA[, c("RNA.1", "RNA.7", "RNA.11", "RNA.17", "RNA.18")])
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Macrophage.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-19.5-macroCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Macrophage Cells at gd 19.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```
NK:
```{r}
avgRNA1 <- avgRNA[, c("RNA.9")]
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Natural.killer.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-19.5-nkCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Natural Killer Cells at gd 19.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```
Smooth muscle:
```{r}
avgRNA1 <- rowMeans(avgRNA[, c("RNA.8", "RNA.26")])
names(avgRNA1) <- rownames(avgRNA)

avgATAC1 <- avgATAC[intersect(rownames(avgATAC), names(avgRNA1)), "RNA.Smooth.muscle.cells"]
names(avgATAC1) <- intersect(rownames(avgATAC), names(avgRNA1))
avgATAC1 <- avgATAC1[order(names(avgATAC1))]
avgRNA1 <- avgRNA1[intersect(rownames(avgATAC), names(avgRNA1))]
avgRNA1 <- avgRNA1[order(names(avgRNA1))]

cor.test(avgRNA1, avgATAC1, method = "spearman", exact = F)

df <- data.frame(avgRNA1=avgRNA1, avgATAC1=avgATAC1)

#pdf("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/FIGURES/20220829_draftV3/figureS4-19.5-smCorr.#pdf", height = 5, width = 6)
ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Smooth Muscle Cells at gd 19.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")
#dev.off()
```

```{r}
sessionInfo()
```
