---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 15,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>

```{r}
library("org.Rn.eg.db")
library("clusterProfiler")

go <- function(genes) {
  GO <- enrichGO(gene = genes, OrgDb=org.Rn.eg.db, ont = "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.05, maxGSSize=500, readable = T, keyType = "ENSEMBL")
  goSimplify <- simplify(GO)
  res <- as.data.frame(goSimplify)
  #res$Rank <- seq(1, nrow(res))
  res$GeneRatio <- sapply(strsplit(res$GeneRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$BgRatio <- sapply(strsplit(res$BgRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$Fold <- as.numeric(res$GeneRatio)/as.numeric(res$BgRatio)
  res <- res[res$qvalue <= 0.05 & res$Fold >= 2 & res$Count >= 5,]
  
  return(res)
}
```

```{r}
library(TissueEnrich)
library(ggplot2)

load(file = "Files/combine-test-expression1.Rdata")

ratHumanOrthologs<-dataset$GRCH38$ratHumanOrthologs
humanGeneMapping<-dataset$GRCH38$humanGeneMapping

#For VentoTormo Data
d <- dataset$PlacentaDeciduaBloodData
data <- d$expressionData
cellDetails <- d$cellDetails

expressionData <- data[intersect(row.names(data), ratHumanOrthologs$Human.gene.stable.ID),]
se <- SummarizedExperiment(assays = SimpleList(as.matrix(expressionData)), rowData = row.names(expressionData), colData = colnames(expressionData))
cellSpecificGenesExp <- teGeneRetrieval(se, expressedGeneThreshold = 1)
```

```{r}
runPCE <- function(inputGenes, nameType = c("Gene.name", "Gene.stable.ID")) {
  if (nameType == "Gene.name") {
    humanOrthologs <- ratHumanOrthologs[ratHumanOrthologs$Gene.name %in% toupper(inputGenes),]
  } else {
    humanOrthologs <- ratHumanOrthologs[ratHumanOrthologs$Gene.stable.ID %in% inputGenes,]
  }
  inputGenes <- humanOrthologs$Human.gene.stable.ID
  print(paste("Number of input genes found is", length(inputGenes)))
  gs <- GeneSet(geneIds=toupper(inputGenes))
  output2 <- teEnrichmentCustom(gs, cellSpecificGenesExp)
  enrichmentOutput <- setNames(data.frame(assay(output2[[1]]), row.names = rowData(output2[[1]])[,1]), colData(output2[[1]])[,1])
  row.names(cellDetails) <- cellDetails$RName
  enrichmentOutput$Tissue <- cellDetails[row.names(enrichmentOutput), "CellName"]
  
  return(enrichmentOutput)
}

plotPCE <- function(enrichmentOutput, name) {
  p <- ggplot(enrichmentOutput,aes(x=reorder(Tissue,-Log10PValue),y=Log10PValue,
                                 label = Tissue.Specific.Genes, fill=Tissue))+
  geom_bar(stat = 'identity')+
    labs(x='', y = '-LOG10(Adj. P-Value)')+
    theme_bw()+
    theme(legend.position='none', plot.margin=unit(c(1,1,1,1),"cm"))+
    theme(plot.title = element_text(hjust = 0.5,size = 20),axis.title =
            element_text(size=15))+
    theme(axis.text.x = element_text(angle = 90, vjust = 1, hjust = 1),
          panel.grid.major= element_blank(),panel.grid.minor = element_blank())+
    ggtitle(name)
  
  return(p)
}
```

TissueEnrich
```{r}
runTE <- function(inputGenes, nameType = c("Gene.name", "Gene.stable.ID")) {
  if (nameType == "Gene.name") {
    humanOrthologs <- ratHumanOrthologs[ratHumanOrthologs$Gene.name %in% toupper(inputGenes),]
  } else {
    humanOrthologs <- ratHumanOrthologs[ratHumanOrthologs$Gene.stable.ID %in% inputGenes,]
  }
  inputGenes <- humanOrthologs$Human.gene.name
  print(paste("Number of input genes found is", length(inputGenes)))
  gs <- GeneSet(geneIds=toupper(inputGenes),organism="Homo Sapiens",geneIdType=SymbolIdentifier())
  output <- teEnrichment(gs)
  enrichmentOutput <- setNames(data.frame(assay(output[[1]]), row.names = rowData(output[[1]])[,1]), colData(output[[1]])[,1])
  enrichmentOutput$Tissue <- rownames(enrichmentOutput)
  
  return(enrichmentOutput)
}
```

## GD 15.5:
```{r}
library(Signac)
library(Seurat)
library(ggplot2)
library(GenomicRanges)
library(patchwork)
set.seed(1234)

closest_genes <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc-toGenes.txt", sep = "\t")
t <- table(closest_genes$gene_name)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5_res0.8.rda")
data

exp <- AverageExpression(data)
exp <- as.data.frame(exp$RNA)
```

```{r}
genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "23")]
colnames(df) <- c("gene", "value")
df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"], alternative = "greater")

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_violin() + ggtitle("Cutoff = 3")
```

```{r}
t2 <- closest_genes[closest_genes$gene_name %in% df[df$category == ">= 3 peaks", "gene"],]
#go(unique(t2$ensembl_gene_id)) #no terms enriched

gd15.5tbc <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scRNA-seq-analysis/3_cluster_DEG/rmChrMTgenes/GD15.5/res0.8/gd15.5-wilcoxin-cluster23.txt", header = T)
length(intersect(df[df$category == ">= 3 peaks", "gene"], gd15.5tbc$gene))
length(unique(df[df$category == ">= 3 peaks", "gene"]))
intersect(df[df$category == ">= 3 peaks", "gene"], gd15.5tbc$gene) #57 genes out of 255 genes are TBC genes
genes <- names(t[t >= 4])
length(intersect(genes, gd15.5tbc$gene)) #25 out of 57 genes have >= 4 peaks
genes <- names(t[t >= 5])
length(genes)
length(intersect(genes, gd15.5tbc$gene)) #16 out of 57 genes have >= 5 peaks
```


## GD 19.5:
```{r}
set.seed(1234)

closest_genes <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc-toGenes.txt", sep = "\t")
t <- table(closest_genes$gene_name)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-4-5-6-7_res0.8.rda")
data2

exp <- AverageExpression(data2)
exp <- as.data.frame(exp$RNA)
```

```{r}
genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "6")]
colnames(df) <- c("gene", "value")
df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"], alternative = "greater")
ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_violin() + ggtitle("Cutoff = 3")
```
```{r}
t2 <- closest_genes[closest_genes$gene_name %in% df[df$category == ">= 3 peaks", "gene"],]
#go(unique(t2$ensembl_gene_id)) #no terms enriched

gd19.5tbc <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scRNA-seq-analysis/3_cluster_DEG/rmChrMTgenes/GD19.5/res0.8/gd19.5-wilcoxin-cluster6.txt", header = T)
length(intersect(df[df$category == ">= 3 peaks", "gene"], gd19.5tbc$gene))
intersect(df[df$category == ">= 3 peaks", "gene"], gd19.5tbc$gene) #39 genes out of 100 genes are TBC genes
genes <- names(t[t >= 4])
length(intersect(genes, gd19.5tbc$gene)) #12 out of 39 genes have >= 4 peaks
genes <- names(t[t >= 5])
length(intersect(genes, gd19.5tbc$gene)) #5 out of 39 genes have >= 5 peaks
```
