---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 15,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>

```{r}
library("org.Rn.eg.db")
library("clusterProfiler")

go <- function(genes) {
  GO <- enrichGO(gene = genes, OrgDb=org.Rn.eg.db, ont = "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.05, maxGSSize=500, readable = T, keyType = "ENSEMBL")
  goSimplify <- simplify(GO)
  res <- as.data.frame(goSimplify)
  #res$Rank <- seq(1, nrow(res))
  res$GeneRatio <- sapply(strsplit(res$GeneRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$BgRatio <- sapply(strsplit(res$BgRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$Fold <- as.numeric(res$GeneRatio)/as.numeric(res$BgRatio)
  res <- res[res$qvalue <= 0.05 & res$Fold >= 2 & res$Count >= 5,]
  
  return(res)
}
```

# gd 15.5
```{r}
library(Signac)
library(Seurat)
library(ggplot2)
library(ggpubr)
library(GenomicRanges)
library(patchwork)
library(dplyr)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5_res0.8.rda")

exp <- AverageExpression(data)
exp <- as.data.frame(exp$RNA)
avgRNA <- exp

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
```

Normalize by peak length before normalizing by default in Seurat:
```{r}
# first compute the length of each peak
rats <- RegionStats(
  object = rats,
  genome = BSgenome.Rnorvegicus.Ensembl.rn6,
  sep = c(":", "-")
)

# extract the width of each peak
peak.width <- GetAssayData(rats, assay = 'MACSpeaks', slot = 'meta.features')$sequence.length

# extract the raw counts
counts <- GetAssayData(rats, assay = 'MACSpeaks', slot = 'counts')

# divide the counts for each peak by the width of the peak
norm.counts <- counts / peak.width

# create a new assay with the normalized counts and perform the same processing 
rats[['normPeaks']] <- CreateAssayObject(counts = norm.counts)
DefaultAssay(rats) <- 'normPeaks'
rats <- RunTFIDF(rats)
rats <- FindTopFeatures(rats)
rats <- RunSVD(rats, n = 50, reduction.key = 'LSI_', reduction.name = 'lsiNorm')
rats <- RunUMAP(rats, dims = 2:20, reduction = 'lsiNorm', reduction.name = 'umapNorm')
avgATAC <- AverageExpression(rats, "normPeaks")
avgATAC <- avgATAC$normPeaks
```
Use all TBC (specific or not) peaks together, genes are markers or not:
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

DefaultAssay(rats) <- "MACSpeaks"
closest_genes <- ClosestFeature(rats, regions = rownames(atac3))
df <- closest_genes

avgRNA1 <- avgRNA$`23`
avgRNA1 <- as.data.frame(avgRNA1)
rownames(avgRNA1) <- rownames(avgRNA)
avgRNA1$genes <- rownames(avgRNA)

avgATAC1 <- avgATAC[rownames(atac3), "Invasive trophoblast cells"]
avgATAC1 <- as.data.frame(avgATAC1)
rownames(avgATAC1) <- rownames(atac3)
avgATAC1$regions <- rownames(atac3)

df <- inner_join(closest_genes, avgATAC1, by = c("query_region" = "regions"))
df <- inner_join(df, avgRNA1, by = c("gene_name" = "genes"))

cor.test(df$avgRNA1, df$avgATAC1, method = "spearman", exact = F)

ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Invasive Trophoblast Cells at gd 15.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")

specificPeaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc-toGenes.txt", sep = "\t")
df2 <- df[df$query_region %in% specificPeaks$query_region,]
cor.test(df2$avgRNA1, df2$avgATAC1, method = "spearman", exact = F)

marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)
df3 <- df2[df2$gene_name %in% marker$gene,]
cor.test(df3$avgRNA1, df3$avgATAC1, method = "spearman", exact = F)

```

```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

DefaultAssay(rats) <- "MACSpeaks"
closest_genes <- ClosestFeature(rats, regions = rownames(atac3))

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "23")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, all peaks all genes GD15.5")

print("Number of genes with >= 3 peaks:")
length(genes)
print("Number of genes with < 3 peaks:")
length(unique(df[df$category == "< 3 peaks", "gene"]))
```

Focus on marker genes:
```{r}
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)
```
Peak distribution of marker genes:
```{r}
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)
t2 <- t[intersect(names(t),marker$gene)]
non <- setdiff(marker$gene, names(t))
nont <- rep(0, length(non))
names(nont) <- non
t3 <- as.vector(t2)
names(t3) <- names(t2)
t3 <- c(t3, nont)
hist(t3, breaks = seq(0, 20, 1))

length(t3[t3 >= 3])
```
202 out of 394 marker genes have >= 3 peaks. <br>

Of the markers with peaks, how many peaks are specific, and is this number significant:
```{r}
specificPeaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc-toGenes.txt", sep = "\t")
df2 <- closest_genes[closest_genes$gene_name %in% marker$gene,]

#number of specific peaks associated to markers
length(intersect(df2$query_region, specificPeaks$query_region))

#number of peaks associated to markers
length(df2$query_region)

#number of non-specific peaks
length(setdiff(closest_genes$query_region, specificPeaks$query_region))

#number of non-specific peaks that associated with markers
temp <- closest_genes[!(closest_genes$query_region %in% specificPeaks$query_region) &
                (closest_genes$gene_name %in% marker$gene),]
dim(temp)

dat <- data.frame(marker=c(423,1345-423), nonmarker=c(3842-423,18760-(3842-423)))
rownames(dat) <- c("specific", "nonspecific")
dat
fisher.test(dat, alternative = "greater")

x <- c()
for (row in rownames(dat)) {
  for (col in colnames(dat)) {
    x <- rbind(x, matrix(rep(c(row, col), dat[row, col]), ncol = 2, byrow = TRUE))
  }
}
df <- as.data.frame(x)
colnames(df) <- c("SpecificPeaks", "MarkerGene")

test <- fisher.test(table(df))

# combine plot and statistical test with ggbarstats
library(ggstatsplot)
ggbarstats(
  df, MarkerGene, SpecificPeaks,
  results.subtitle = FALSE,
  proportion.test=F,
  subtitle = paste0(
    "Fisher's exact test", ", p-value = ",
    ifelse(test$p.value < 0.001, "< 0.001", round(test$p.value, 3))
  )
)
```

Markers that only have nonspecific peaks:
```{r}
nonSpe <- df2[df2$query_region %in% setdiff(closest_genes$query_region, specificPeaks$query_region),]
onlyNonSpe <- nonSpe[!(nonSpe$gene_name %in% specificPeaks$gene_name),]

go_onlyNonSpe <- go(rnor6[rnor6$gene_name %in% onlyNonSpe$gene_name, 1])
write.table(go_onlyNonSpe, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-allTBCpeaks-go_genesWithOnlyNonSpePeaks.txt", sep = "\t", quote = F, row.names = F)

r<-go_onlyNonSpe[grep("epith|placenta|angiogenesis|prolactin|cell migration|cell motility", go_onlyNonSpe$Description),]

ggplot(data=r, aes(x=reorder(Description, -log10(as.numeric(qvalue))), y=-log10(as.numeric(qvalue)))) +
  geom_bar(stat="identity", fill="#008837", width = 0.7) + coord_flip() +
  labs(title = "Markers with only nonspecific peaks",
       y = "-log10(q-value)",
       x = "") +
  theme(plot.title = element_text(size = 25, face = "bold"), legend.text=element_text(size=25),
        axis.text=element_text(size=25), axis.title.x = element_text(size = 25))
```

Markers that only have specific peaks:
```{r}
onlySpe <- specificPeaks[!(specificPeaks$gene_name %in% nonSpe$gene_name),]
go_onlySpe <- go(rnor6[rnor6$gene_name %in% onlySpe$gene_name, 1])

write.table(go_onlySpe, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-allTBCpeaks-go_genesWithOnlySpePeaks.txt", sep = "\t", quote = F, row.names = F)


r<-go_onlySpe[grep("epith|placenta|angiogenesis|prolactin|cell migration|cell motility", go_onlySpe$Description),]

ggplot(data=r, aes(x=reorder(Description, -log10(as.numeric(qvalue))), y=-log10(as.numeric(qvalue)))) +
  geom_bar(stat="identity", fill="#008837", width = 0.7) + coord_flip() +
  labs(title = "Markers with only specific peaks",
       y = "-log10(q-value)",
       x = "") +
  theme(plot.title = element_text(size = 25, face = "bold"), legend.text=element_text(size=25),
        axis.text=element_text(size=25), axis.title.x = element_text(size = 25))
```

Markers with both:
```{r}
both <- specificPeaks[specificPeaks$gene_name %in% intersect(specificPeaks$gene_name, nonSpe$gene_name),]
go_both <- go(rnor6[rnor6$gene_name %in% both$gene_name, 1])

write.table(go_both, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-allTBCpeaks-go_genesWithBothPeaks.txt", sep = "\t", quote = F, row.names = F)

r<-go_both[grep("epith|placenta|angiogenesis|prolactin|cell migration|cell motility", go_both$Description),]

ggplot(data=r, aes(x=reorder(Description, -log10(as.numeric(qvalue))), y=-log10(as.numeric(qvalue)))) +
  geom_bar(stat="identity", fill="#008837", width = 0.7) + coord_flip() +
  labs(title = "Markers with both types of peaks",
       y = "-log10(q-value)",
       x = "") +
  theme(plot.title = element_text(size = 25, face = "bold"), legend.text=element_text(size=25),
        axis.text=element_text(size=25), axis.title.x = element_text(size = 25))
```

## GD 19.5:

```{r}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-4-5-6-7_res0.8.rda")
data2

exp <- AverageExpression(data2)
exp <- as.data.frame(exp$RNA)
avgRNA <- exp

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
```
Use all TBC (specific or not) peaks together, genes are markers or not:
Normalize by peak length before normalizing by default in Seurat:
```{r}
# first compute the length of each peak
rats <- RegionStats(
  object = rats,
  genome = BSgenome.Rnorvegicus.Ensembl.rn6,
  sep = c(":", "-")
)

# extract the width of each peak
peak.width <- GetAssayData(rats, assay = 'MACSpeaks', slot = 'meta.features')$sequence.length

# extract the raw counts
counts <- GetAssayData(rats, assay = 'MACSpeaks', slot = 'counts')

# divide the counts for each peak by the width of the peak
norm.counts <- counts / peak.width

# create a new assay with the normalized counts and perform the same processing 
rats[['normPeaks']] <- CreateAssayObject(counts = norm.counts)
DefaultAssay(rats) <- 'normPeaks'
rats <- RunTFIDF(rats)
rats <- FindTopFeatures(rats)
rats <- RunSVD(rats, n = 50, reduction.key = 'LSI_', reduction.name = 'lsiNorm')
rats <- RunUMAP(rats, dims = 2:20, reduction = 'lsiNorm', reduction.name = 'umapNorm')
avgATAC <- AverageExpression(rats, "normPeaks")
avgATAC <- avgATAC$normPeaks
```
Use all TBC (specific or not) peaks together, genes are markers or not:
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

DefaultAssay(rats) <- "MACSpeaks"
closest_genes <- ClosestFeature(rats, regions = rownames(atac3))
df <- closest_genes

avgRNA1 <- avgRNA$`6`
avgRNA1 <- as.data.frame(avgRNA1)
rownames(avgRNA1) <- rownames(avgRNA)
avgRNA1$genes <- rownames(avgRNA)

avgATAC1 <- avgATAC[rownames(atac3), "Invasive trophoblast cells"]
avgATAC1 <- as.data.frame(avgATAC1)
rownames(avgATAC1) <- rownames(atac3)
avgATAC1$regions <- rownames(atac3)

df <- inner_join(closest_genes, avgATAC1, by = c("query_region" = "regions"))
df <- inner_join(df, avgRNA1, by = c("gene_name" = "genes"))

cor.test(df$avgRNA1, df$avgATAC1, method = "spearman", exact = F)

ggplot(df, aes(x=log10(avgRNA1+10^(-5)), y=log10(avgATAC1+10^(-5)))) + geom_point(alpha = 1/10) + geom_smooth(method=lm, se=FALSE, fullrange=TRUE) + ggtitle("Correlation in Invasive Trophoblast Cells at gd 19.5") + stat_cor(method = "spearman", label.y = -4, cor.coef.name = "rho")

specificPeaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc-toGenes.txt", sep = "\t")
df2 <- df[df$query_region %in% specificPeaks$query_region,]
cor.test(df2$avgRNA1, df2$avgATAC1, method = "spearman", exact = F)

marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T)
df3 <- df2[df2$gene_name %in% marker$gene,]
cor.test(df3$avgRNA1, df3$avgATAC1, method = "spearman", exact = F)

```

```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

DefaultAssay(rats) <- "MACSpeaks"
closest_genes <- ClosestFeature(rats, regions = rownames(atac3))

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "6")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, all peaks all genes GD19.5")

print("Number of genes with >= 3 peaks:")
length(genes)
print("Number of genes with < 3 peaks:")
length(unique(df[df$category == "< 3 peaks", "gene"]))
```

Focus on marker genes:
```{r}
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T)
```
Peak distribution of marker genes:
```{r}
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)
t2 <- t[intersect(names(t),marker$gene)]
non <- setdiff(marker$gene, names(t))
nont <- rep(0, length(non))
names(nont) <- non
t3 <- as.vector(t2)
names(t3) <- names(t2)
t3 <- c(t3, nont)
hist(t3, breaks = seq(0, 20, 1))

length(t3[t3 >= 3])
```
202 out of 394 marker genes have >= 3 peaks. <br>

Of the markers with peaks, how many peaks are specific
```{r}
specificPeaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc-toGenes.txt", sep = "\t")
df2 <- closest_genes[closest_genes$gene_name %in% marker$gene,]

#number of specific peaks
length(specificPeaks$query_region)

#number of specific peaks associated to markers
length(intersect(df2$query_region, specificPeaks$query_region))

#number of peaks associated to markers
length(df2$query_region)

#number of non-specific peaks
length(setdiff(closest_genes$query_region, specificPeaks$query_region))

#number of non-specific peaks that associated with markers
temp <- closest_genes[!(closest_genes$query_region %in% specificPeaks$query_region) &
                (closest_genes$gene_name %in% marker$gene),]
dim(temp)

dat <- data.frame(marker=c(339,765-339), nonmarker=c(2033-339,7808-(2033-339)))
rownames(dat) <- c("specific", "nonspecific")
dat
fisher.test(dat, alternative = "greater")

x <- c()
for (row in rownames(dat)) {
  for (col in colnames(dat)) {
    x <- rbind(x, matrix(rep(c(row, col), dat[row, col]), ncol = 2, byrow = TRUE))
  }
}
df <- as.data.frame(x)
colnames(df) <- c("SpecificPeaks", "MarkerGene")

test <- fisher.test(table(df))

# combine plot and statistical test with ggbarstats
library(ggstatsplot)
ggbarstats(
  df, MarkerGene, SpecificPeaks,
  results.subtitle = FALSE,
  proportion.test=F,
  subtitle = paste0(
    "Fisher's exact test", ", p-value = ",
    ifelse(test$p.value < 0.001, "< 0.001", round(test$p.value, 3))
  )
)

```

Markers that only have nonspecific peaks:
```{r}
nonSpe <- df2[df2$query_region %in% setdiff(closest_genes$query_region, specificPeaks$query_region),]
onlyNonSpe <- nonSpe[!(nonSpe$gene_name %in% specificPeaks$gene_name),]

go_onlyNonSpe <- go(rnor6[rnor6$gene_name %in% onlyNonSpe$gene_name, 1])
write.table(go_onlyNonSpe, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-allTBCpeaks-go_genesWithOnlyNonSpePeaks.txt", sep = "\t", quote = F, row.names = F)

r<-go_onlyNonSpe[grep("epith|placenta|angiogenesis|prolactin|cell migration|cell motility", go_onlyNonSpe$Description),]

ggplot(data=r, aes(x=reorder(Description, -log10(as.numeric(qvalue))), y=-log10(as.numeric(qvalue)))) +
  geom_bar(stat="identity", fill="#008837", width = 0.7) + coord_flip() +
  labs(title = "Markers with only nonspecific peaks",
       y = "-log10(q-value)",
       x = "") +
  theme(plot.title = element_text(size = 25, face = "bold"), legend.text=element_text(size=25),
        axis.text=element_text(size=25), axis.title.x = element_text(size = 25))
```

Markers that only have specific peaks:
```{r}
onlySpe <- specificPeaks[!(specificPeaks$gene_name %in% nonSpe$gene_name),]
go_onlySpe <- go(rnor6[rnor6$gene_name %in% onlySpe$gene_name, 1])

write.table(go_onlySpe, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-allTBCpeaks-go_genesWithOnlySpePeaks.txt", sep = "\t", quote = F, row.names = F)

r<-go_onlySpe[grep("epith|placenta|angiogenesis|prolactin|cell migration|cell motility", go_onlySpe$Description),]

ggplot(data=r, aes(x=reorder(Description, -log10(as.numeric(qvalue))), y=-log10(as.numeric(qvalue)))) +
  geom_bar(stat="identity", fill="#008837", width = 0.7) + coord_flip() +
  labs(title = "Markers with only specific peaks",
       y = "-log10(q-value)",
       x = "") +
  theme(plot.title = element_text(size = 25, face = "bold"), legend.text=element_text(size=25),
        axis.text=element_text(size=25), axis.title.x = element_text(size = 25))
```

Markers with both:
```{r}
both <- specificPeaks[specificPeaks$gene_name %in% intersect(specificPeaks$gene_name, nonSpe$gene_name),]
go_both <- go(rnor6[rnor6$gene_name %in% both$gene_name, 1])

write.table(go_both, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-allTBCpeaks-go_genesWithBothPeaks.txt", sep = "\t", quote = F, row.names = F)

r<-go_both[grep("epith|placenta|angiogenesis|prolactin|cell migration|cell motility", go_both$Description),]

ggplot(data=r, aes(x=reorder(Description, -log10(as.numeric(qvalue))), y=-log10(as.numeric(qvalue)))) +
  geom_bar(stat="identity", fill="#008837", width = 0.7) + coord_flip() +
  labs(title = "Markers with both types of peaks",
       y = "-log10(q-value)",
       x = "") +
  theme(plot.title = element_text(size = 25, face = "bold"), legend.text=element_text(size=25),
        axis.text=element_text(size=25), axis.title.x = element_text(size = 25))
```


```{r}
sessionInfo()
```
