---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 15,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>
  
Load libraries
```{r}
library(stringr)
library(dplyr)
library(Seurat)
library(Signac)
library(Seurat)
library(S4Vectors)
library(GenomicRanges)
library(patchwork)
library(biomaRt)
library(ggplot2)
set.seed(1234)
```

Load EVT peaks to genes (by GREAT)
```{r}
geneToPeak <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/EVT_GREAT/EVT_genesToPeaks.txt", header = F, sep = "\t", fill = T, stringsAsFactors = F, quote = "")
geneToPeak$noPeaks <- str_count(geneToPeak$V2, "EVT_peaks")
quantile(geneToPeak$noPeaks)

conservedRats <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/conservedRat.txt", header = T)

load(file = "/work/LAS/geetu-lab/hhvu/combine-test-expression1.Rdata")
ratHumanOrthologs <- dataset$GRCH38$ratHumanOrthologs
ratHumanOrthologs <- ratHumanOrthologs[ratHumanOrthologs$Gene.name %in% conservedRats$ratGenes,]

geneToPeak2 <- geneToPeak[geneToPeak$V1 %in% ratHumanOrthologs$Human.gene.name,]
temp <- data_frame(V1=setdiff(ratHumanOrthologs$Human.gene.name, geneToPeak2$V1), V2="none", noPeaks=0)
geneToPeak2 <- rbind(geneToPeak2, temp)
quantile(geneToPeak2$noPeaks)
```

# Number of EVT peaks that the genes with >= 3 peaks in rat have:
# gd 15.5
```{r}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges

c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

closest_genes <- ClosestFeature(rats, regions = rownames(atac3))

closest_genes$gene_name <- toupper(closest_genes$gene_name)

load(file = "/work/LAS/geetu-lab/hhvu/combine-test-expression1.Rdata")
ratHumanOrthologs <- dataset$GRCH38$ratHumanOrthologs
closest_genes2 <- inner_join(closest_genes, ratHumanOrthologs[,c(2,4)], by = c("gene_name" = "Gene.name"))
t <- table(closest_genes2$Human.gene.name)
t <- as.data.frame(t)
t$Var1 <- toupper(t$Var1)

geneToPeak3 <- geneToPeak[geneToPeak$V1 %in% closest_genes2$Human.gene.name,]
geneToPeak3 <- inner_join(geneToPeak3, t, by = c("V1" = "Var1"))
geneToPeak3$category <- ifelse(geneToPeak3$Freq >= 3, ">= 3 rats", "< 3 rats")
wilcox.test(geneToPeak3[geneToPeak3$category == ">= 3 rats", "noPeaks"], geneToPeak3[geneToPeak3$category == "< 3 rats", "noPeaks"])

ggplot(geneToPeak3, aes(x=category, y=noPeaks)) + geom_boxplot() + ggtitle("Cutoff = 3, invasive trophoblast GD15.5")

s <- geneToPeak3[,1:4]
colnames(s) <- c("humanGene", "EVTpeaks", "noEVTpeaks", "noRatPeaks")
write.table(s, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/EVT_GREAT/gd15.5EVTgenesWithNonspeAndSpeRatPeaks.txt", sep = "\t", quote = F, row.names = F)
```

# gd 19.5
```{r}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges

c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

closest_genes <- ClosestFeature(rats, regions = rownames(atac3))
#write.table(closest_genes, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-allTBCpeaks-toGenes.txt", sep = "\t", quote = F, row.names = F)
closest_genes$gene_name <- toupper(closest_genes$gene_name)

load(file = "/work/LAS/geetu-lab/hhvu/combine-test-expression1.Rdata")
ratHumanOrthologs <- dataset$GRCH38$ratHumanOrthologs
closest_genes2 <- inner_join(closest_genes, ratHumanOrthologs[,c(2,4)], by = c("gene_name" = "Gene.name"))
t <- table(closest_genes2$Human.gene.name)
t <- as.data.frame(t)
t$Var1 <- toupper(t$Var1)

geneToPeak3 <- geneToPeak[geneToPeak$V1 %in% closest_genes2$Human.gene.name,]
geneToPeak3 <- inner_join(geneToPeak3, t, by = c("V1" = "Var1"))
geneToPeak3$category <- ifelse(geneToPeak3$Freq >= 3, ">= 3 rats", "< 3 rats")
wilcox.test(geneToPeak3[geneToPeak3$category == ">= 3 rats", "noPeaks"], geneToPeak3[geneToPeak3$category == "< 3 rats", "noPeaks"])

ggplot(geneToPeak3, aes(x=category, y=noPeaks)) + geom_boxplot() + ggtitle("Cutoff = 3, invasive trophoblast GD19.5")

s <- geneToPeak3[,1:4]
colnames(s) <- c("humanGene", "EVTpeaks", "noEVTpeaks", "noRatPeaks")
write.table(s, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/EVT_GREAT/gd19.5EVTgenesWithNonspeAndSpeRatPeaks.txt", sep = "\t", quote = F, row.names = F)
```