---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 15,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>

We first define common peaks to regions enriched in TBC cluster at GD15.5 and GD19.5 that overlap by at least 50\%. <br>
```{BASH, eval = F}
module load bedtools2

bedtools intersect -a /work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc.bed -b /work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc.bed -f 0.5 -r -wa -wb > /work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks.bed

cd /work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/
cut -f1-4 commonPeaks.bed | sort -u > commonPeaks-gd15.5.bed
cut -f5-8 commonPeaks.bed | sort -u > commonPeaks-gd19.5.bed
```

From hereafter, we will be using GD19.5 coordinates of the common peaks for downstream analyses. <br>

The distribution of the common peaks in the genome using GD19.5 coordinates:
```{r}
library(ChIPseeker)
library(GenomicFeatures)
library(RMariaDB) #remember to module load mariadb/10.1.23-py2-mpich-qfxutlj

txdb <- makeTxDbFromEnsembl(organism= "Rattus norvegicus", release = 98)

file <- "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks-gd19.5.bed"
peakAnno <- annotatePeak(file, tssRegion=c(-3000, 3000), TxDb=txdb)

plotDistToTSS(peakAnno,
              title="Distribution of transcription factor-binding loci relative to TSS")
```

Using GD19.5 coordinates, associate the common peaks to the nearest genes, then run GO analysis.
```{r}
library(Signac)
library(Seurat)
library(JASPAR2020)
library(TFBSTools)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
library(patchwork)
library(dplyr)
library(org.Rn.eg.db)
library(clusterProfiler)
set.seed(1234)

go <- function(genes) {
  GO <- enrichGO(gene = genes, OrgDb=org.Rn.eg.db, ont = "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.05, maxGSSize=500, readable = T, keyType = "ENSEMBL")
  goSimplify <- simplify(GO)
  res <- as.data.frame(goSimplify)
  #res$Rank <- seq(1, nrow(res))
  res$GeneRatio <- sapply(strsplit(res$GeneRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$BgRatio <- sapply(strsplit(res$BgRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$Fold <- as.numeric(res$GeneRatio)/as.numeric(res$BgRatio)
  res <- res[res$qvalue <= 0.05 & res$Fold >= 2 & res$Count >= 5,]
  
  return(res)
}

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
peaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks-gd19.5.bed", header=F)
closest_genes <- ClosestFeature(rats, regions = peaks$V4)

r <- go(closest_genes$ensembl_gene_id)
head(r)
```

## Motif analysis:
We carry out motif analysis with TF motifs from rat, mouse and human database. Firstly, run with rat TFs:
```{r, eval = F}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'

peaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks.bed", header=F)
top.peaks <- unique(peaks$V8)

open.peaks <- AccessiblePeaks(rats, assay = "MACSpeaks", min.cells = 1)
meta.feature <- GetAssayData(rats, assay = "MACSpeaks", slot = "meta.features")
peaks.matched <- MatchRegionStats(meta.feature = meta.feature[open.peaks, ], query.feature = meta.feature[top.peaks, ], n = 50000, features.match = c("GC.percent", "sequence.length"))

#rats
pfm <- getMatrixSet(x = JASPAR2020, opts = list(species = "Rattus norvegicus", all_versions = FALSE))
motif.matrix <- CreateMotifMatrix(features = granges(rats), pwm = pfm, genome = BSgenome.Rnorvegicus.Ensembl.rn6, p.cutoff = 0.00005)
motif.positions <- motifmatchr::matchMotifs(pwms = pfm, subject = granges(rats), out = 'positions', genome = BSgenome.Rnorvegicus.Ensembl.rn6)
motif <- CreateMotifObject(data = motif.matrix, positions = motif.positions, pwm = pfm)
rats <- SetAssayData(object = rats, slot = 'motifs', new.data = motif)
rats <- RegionStats(rats[['MACSpeaks']], genome=BSgenome.Rnorvegicus.Ensembl.rn6)

enriched.motifs <- FindMotifs(object=rats, features=top.peaks, background=peaks.matched)

#write.table(enriched.motifs, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/common19.5-ratsMotifs.txt", quote=F)
#save(rats, file = "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/gd19.5_ATACwithRatPFM.rda")
```
Mouse motifs:
```{r, eval = F}
pfm <- getMatrixSet(x = JASPAR2020, opts = list(species = "Mus musculus", all_versions = FALSE))
motif.matrix <- CreateMotifMatrix(features = granges(rats), pwm = pfm, genome = BSgenome.Rnorvegicus.Ensembl.rn6, p.cutoff = 0.00005)
motif.positions <- motifmatchr::matchMotifs(pwms = pfm, subject = granges(rats), out = 'positions', genome = BSgenome.Rnorvegicus.Ensembl.rn6)
motif <- CreateMotifObject(data = motif.matrix, positions = motif.positions, pwm = pfm)
rats <- SetAssayData(object = rats, slot = 'motifs', new.data = motif)

enriched.motifs <- FindMotifs(object=rats, features=top.peaks, background=peaks.matched)

#write.table(enriched.motifs, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/common19.5-mouseMotifs.txt", quote=F)
#save(rats, file = "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/gd19.5_ATACwithMousePFM.rda")
```
Human motifs:
```{r, eval = F}
pfm <- getMatrixSet(x = JASPAR2020, opts = list(species = "Homo sapiens", all_versions = FALSE))
motif.matrix <- CreateMotifMatrix(features = granges(rats), pwm = pfm, genome = BSgenome.Rnorvegicus.Ensembl.rn6, p.cutoff = 0.00005)
motif.positions <- motifmatchr::matchMotifs(pwms = pfm, subject = granges(rats), out = 'positions', genome = BSgenome.Rnorvegicus.Ensembl.rn6)
motif <- CreateMotifObject(data = motif.matrix, positions = motif.positions, pwm = pfm)
rats <- SetAssayData(object = rats, slot = 'motifs', new.data = motif)

enriched.motifs <- FindMotifs(object=rats, features=top.peaks, background=peaks.matched)

#write.table(enriched.motifs, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/common19.5-humanMotifs.txt", quote=F)
#save(rats, file = "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/gd19.5_ATACwithHumanPFM.rda")
```

Next, we compile the enriched motifs to one table.
```{r}
rat19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/common19.5-ratsMotifs.txt", header = T)
rat19.5$adj_pval <- p.adjust(rat19.5$pvalue, "BH")
rat19.5 <- rat19.5[rat19.5$adj_pval <= 0.05 & rat19.5$fold.enrichment >= 1.5,]
rat19.5$type <- "rat"

mouse19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/common19.5-mouseMotifs.txt", header = T)
mouse19.5$adj_pval <- p.adjust(mouse19.5$pvalue, "BH")
mouse19.5 <- mouse19.5[mouse19.5$adj_pval <= 0.05 & mouse19.5$fold.enrichment >= 1.5,]
mouse19.5$type <- "mouse"

human19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/common19.5-humanMotifs.txt", header = T)
human19.5$adj_pval <- p.adjust(human19.5$pvalue, "BH")
human19.5 <- human19.5[human19.5$adj_pval <= 0.05 & human19.5$fold.enrichment >= 1.5,]
human19.5$type <- "human"

motifs19.5 <- rbind(rat19.5, mouse19.5, human19.5)

dim(motifs19.5)
```

Next, filter and keep only transcription factors (TFs) with enriched motifs and have expression levels > 1 at GD19.5.
```{r}
mouseOrthologs <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/rat-mouseOrthologs.txt", header = T, sep = "\t")
mouseOrthologs <- as.data.frame(apply(mouseOrthologs, 2, toupper))
humanOrthologs <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/rat-humanOrthologs.txt", header = T, sep = "\t")
humanOrthologs <- as.data.frame(apply(humanOrthologs, 2, toupper))

#transforming functions
subVar2 <- function(x) {
  ifelse(grepl("var.2", x, ignore.case = TRUE), gsub("\\(var.2\\)", "", x, ignore.case = TRUE), x)
}
subVar3 <- function(x) {
  ifelse(grepl("var.3", x, ignore.case = TRUE), gsub("\\(var.3\\)", "", x, ignore.case = TRUE), x)
}
subColon <- function(x) {
  ifelse(grepl("::", x, ignore.case = TRUE), strsplit(x, "::"), x)
}

enriched <- motifs19.5[, c("motif", "motif.name", "type")]
enriched2 <- data.frame(motif=NA, motif.name=NA, type=NA, TFgenes=NA)
for (i in 1:nrow(enriched)) {
  if (enriched[i, "type"] == "rat") {
    temp <- enriched[i,]
    temp$TFgenes <- enriched[i, "motif.name"]
    enriched2 <- rbind(enriched2, temp)
  } else if (enriched[i, "type"] == "mouse") {
    temp <- enriched[i,]
    names <- sapply(enriched[i, "motif.name"], subVar2)
    names <- sapply(names, subVar3)
    names <- sapply(names, subColon)
    names <- unlist(names)
    df <- data.frame(tf=toupper(names))
    df <- inner_join(df, mouseOrthologs[, c("Gene.name", "Gene.name.1")], by = c("tf" = "Gene.name.1"))
    temp2 <- temp[rep(seq_len(nrow(temp)), each = length(df$Gene.name)), ]
    temp2$TFgenes <- df$Gene.name
    enriched2 <- rbind(enriched2, temp2)
  } else {
    temp <- enriched[i,]
    names <- sapply(enriched[i, "motif.name"], subVar2)
    names <- sapply(names, subVar3)
    names <- sapply(names, subColon)
    names <- unlist(names)
    df <- data.frame(tf=toupper(names))
    df <- inner_join(df, mouseOrthologs[, c("Gene.name", "Gene.name.1")], by = c("tf" = "Gene.name.1"))
    temp2 <- temp[rep(seq_len(nrow(temp)), each = length(df$Gene.name)), ]
    temp2$TFgenes <- df$Gene.name
    enriched2 <- rbind(enriched2, temp2)
  }
}

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-4-5-6-7_res0.8.rda")
exp19.5 <- AverageExpression(data2)
exp19.5 <- as.data.frame(exp19.5$RNA)
rownames(exp19.5) <- toupper(rownames(exp19.5))
g2 <- rownames(exp19.5[rownames(exp19.5) %in% enriched2$TFgenes & exp19.5$`6` > 1,])

enriched2 <- enriched2[enriched2$TFgenes %in% g2,]

motifs19.5 <- motifs19.5[motifs19.5$motif.name %in% enriched2$motif.name,]

dim(motifs19.5)
```
Convert common peaks (GD19.5 coordinates) to human hg38 coordinates with LiftOver, then find motifs with high matching scores with RGT. A matching score is high if it's in 0.9 quantile of all the scores.
```{BASH, eval=F}
cd /work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks

module load python/3.8.0-k7w5uj4
module load py-numpydoc/0.6.0-py2-iady7mn
module load py-scipy/1.6.0-py3-dwtst2n
module load py-cython/0.29.21-py3-yfooekb

rgt-motifanalysis matching --organism hg38 --rand-proportion 10 --fpr 0.001 --input-file commonPeaks-gd15.5-humanLiftOver.bed --motif-dbs /work/LAS/geetu-lab/hhvu/rgtdata/motifs/jaspar_vertebrates/ --output-location /work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/rgt-motifmatch/
```
```{r}
highMotifsHumanSeq <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/rgt-motifmatch/commonPeaks-gd19.5-humanLiftOver_mpbs.bed", header = F, sep = "\t")
highMotifsHumanSeq$ID <- substr(highMotifsHumanSeq$V4, 1, 8)
highMotifsHumanSeq <- highMotifsHumanSeq[highMotifsHumanSeq$V5 >= quantile(highMotifsHumanSeq$V5, 0.9),]
motifs19.5 <- motifs19.5[motifs19.5$motif %in% highMotifsHumanSeq$ID,]
dim(motifs19.5)
```

Find peaks that contain the enriched motifs of the expressed TFs. Since the types of the retained motifs were mouse and human, we will load the related objects only.
```{r}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/gd19.5_ATACwithMousePFM.rda")

temp <- motifs19.5[motifs19.5$type == "mouse",]
motifPeaks <- data.frame(motif=NA, peak=NA)
for (i in 1:nrow(temp)) {
  p <- intersect(closest_genes$query_region, names(which(GetMotifData(rats)[, temp$motif[i]] == 1)))
  temp1 <- data.frame(motif=rep(temp$motif[i], length(p)), 
                      peak=p)
  motifPeaks <- rbind(motifPeaks, temp1)
}

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/GD19.5_motifAnalysis/gd19.5_ATACwithHumanPFM.rda")

temp <- motifs19.5[motifs19.5$type == "human",]
for (i in 1:nrow(temp)) {
  p <- intersect(closest_genes$query_region, names(which(GetMotifData(rats)[, temp$motif[i]] == 1)))
  temp1 <- data.frame(motif=rep(temp$motif[i], length(p)), 
                      peak=p)
  motifPeaks <- rbind(motifPeaks, temp1)
}

motifPeaks <- motifPeaks[!is.na(motifPeaks$motif),]

dim(motifPeaks)
```


## Common peak - common gene networks:
Build list of genes that are TBC markers at both gd15.5 and gd19.5.
```{r}
marker15.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T, sep = "\t")
marker19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T, sep = "\t")
markers <- intersect(intersect(marker15.5$gene, marker19.5$gene), closest_genes$gene_name)
conservedRats <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/conservedRat.txt", header = T, sep = "\t")

markers <- markers[which(toupper(markers) %in% conservedRats$ratGenes == T)]

length(markers)
```
Find common peaks that are in the promoter regions of genes. First we obtain gene promoters (defined to be 1000bp around gene TSS), then take any common peaks (gd19.5 coordinates) that overlap with the promoters.
```{r, eval = F}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.rda")
t <- rnor6[,c("chromosome_name", "start_position", "end_position", "strand", "ensembl_gene_id", "gene_name")]
for (i in 1:dim(t)[1]){
  if (t$strand[i] == "+") {
    t$end_position[i] = t$start_position[i] + 500
    t$start_position[i] = t$start_position[i] - 500
  }
  else if (t$strand[i] == "-") {
    t$start_position[i] = t$end_position[i] - 500
    t$end_position[i] = t$end_position[i] + 500
  }
}
t$chromosome_name <- paste0("chr", t$chromosome_name)
colnames(t)[1:3] <- c("chr", "start", "end")
t <- t[with(t, order(chr, start)),]
#write.table(t, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/rn6.promoters.bed", sep = "\t", quote = F, row.names = F)
```
```{bash, eval=F}
cd /work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/
module load bedtools2 
sed -i 's/chr//g' rn6.promoters.bed
bedtools intersect -wa -wb -a 5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks-gd19.5.bed -b rn6.promoters.bed > 5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks-gd19.5-promoterPeaks.bed
```

Associate peaks/motifs to genes
```{r}
network <- inner_join(closest_genes, motifPeaks, by = c("query_region" = "peak"))
network <- inner_join(network, motifs19.5[,c("motif", "motif.name")], by = c("motif" = "motif"))
network <- network[network$gene_name %in% markers,]
```

Next we keep only distal peaks (peaks that are not promoter peaks) that are co-accessible with promoter peaks. First, using Cicero to calculate the co-accessibility scores between pairs of peaks at GD19.5:
```{r, eval = F}
library(Signac)
library(Seurat)
library(SeuratWrappers)
library(ggplot2)
library(patchwork)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
library(cicero)
set.seed(1234)

size <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/rn6.size", header = F)
size <- size[1:22,]
colnames(size) <- c("chr", "length")

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
rats.cds <- as.cell_data_set(x = rats)
rats.cicero <- make_cicero_cds(rats.cds, reduced_coordinates= reducedDims(rats.cds)$UMAP)
conns19.5 <- run_cicero(rats.cicero, genomic_coords= size, sample_num=100)
#save(conns19.5, file = "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/conns-gd19.5.rda")
```

```{r}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/conns-gd19.5.rda")
promoterPeaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks-gd19.5-promoterPeaks.bed", header = F)

#step 1 - keep the distal - promoter peak pairs that have coaccess > 0
conns19.5 <- conns[(conns$Peak1 %in% promoterPeaks$V4 | 
                      conns$Peak2 %in% promoterPeaks$V4) &
                     conns$coaccess > 0,]

#step 2 - keep genes that have common promoter peaks
network <- network[network$gene_name %in% promoterPeaks$V10,]

network <- network[network$query_region %in% setdiff(c(conns19.5$Peak1, conns19.5$Peak2), promoterPeaks$V4),]
```


Grouping motifs of the same gene families together
```{r}
jaspar <- read.table("/work/LAS/geetu-lab/hhvu/JASPAR-library.txt", header = T, sep = "\t")
jaspar <- jaspar[jaspar$ID %in% network$motif,]
network <- inner_join(network, jaspar[,c("ID", "Family")], by = c("motif" = "ID"))
keep <- network$Family
keep <- subColon(keep)
for (i in 1:length(keep)) {
  keep[i] <- list(unique(keep[i][[1]]))
  for (j in setdiff(1:length(keep), which(keep %in% keep[i]))) {
    if (sum(keep[j][[1]] %in% keep[i][[1]]) > 0) {
      keep[j] <- keep[i]
    }
  }
}
keep2 <- unlist(lapply(keep, toString))
network$Family2 <- keep2

all <- data.frame(tfGene=character(), regGene=character())
for (i in unique(network$Family2)) {
  tfGene <- unique(network[which(network$Family2 == i), "motif.name"])
  tf <- toString(tfGene)
  regGene <- unique(network[which(network$Family2 == i), "gene_name"])
  tbl <- data.frame(matrix(nrow = length(tfGene)*length(regGene), ncol = 2))
  for (j in 1:length(tfGene)) {
    tbl[((j-1)*length(regGene)+1):(j*length(regGene)),1] <- tf
    tbl[((j-1)*length(regGene)+1):(j*length(regGene)),2] <- as.character(regGene)
  }
  all <- rbind(all, tbl)
}
all$X2 <- toupper(all$X2)
all <- dplyr::distinct(all)

g <- unique(all$X2)
for (k in 1:length(g)) {
  if (length(grep(g[k], all$X1)) != 0) {
    all[which(g[k] == all$X2), "X2"] <- all[grep(g[k], all$X1)[1], "X1"]
  }
}
all <- dplyr::distinct(all)

#write.table(all, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/conns-commonPeaksGenes-gd9.5.txt", sep = "\t", row.names = F, quote = F)
```

Build network in Cytoscape.
```{r}
knitr::include_graphics("Files/conns-PeaksGenes-gd19.5.PNG")
```
PKIB https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7782383/

```{r}
library(ensembldb)
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/unintegrated_2timepoints.rda")
unintegrated <- RenameIdents(unintegrated, `GD15.5_TBC` = "GD15.5_Trophoblast_cells", `GD19.5_TBC` = "GD19.5_Trophoblast_cells")
DefaultAssay(unintegrated) <- "MACSpeaks"
file <- ensDbFromGtf("/work/LAS/geetu-lab/hhvu/project3_scATAC/Rattus_norvegicus.Rnor_6.0.98.gtf")
rnor6.ensDB <- EnsDb(file)
annotations <- suppressWarnings(GetGRangesFromEnsDb(ensdb = rnor6.ensDB))
seqlevelsStyle(annotations) <- "Ensembl"
Annotation(unintegrated) <- annotations


cols <- c("#3e5d66", "#3e5d66")

cov_plot <- CoveragePlot(
  object = unintegrated,
  region = "20-40648139-40648778",
  feature = "Pkib",
  annotation = TRUE,
  peaks = TRUE,
  heights = c(6, 1, 1),
  extend.upstream = 5000,
  extend.downstream = 5000,
  idents = c("GD15.5_Trophoblast_cells", "GD19.5_Trophoblast_cells")
)

cov_plot
```


## Do not use conns
Build list of genes that are TBC markers at both gd15.5 and gd19.5.
```{r}
marker15.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T, sep = "\t")
marker19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T, sep = "\t")
markers <- intersect(intersect(marker15.5$gene, marker19.5$gene), closest_genes$gene_name)
conservedRats <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/conservedRat.txt", header = T, sep = "\t")

markers <- markers[which(toupper(markers) %in% conservedRats$ratGenes == T)]

length(markers)
```

Associate peaks/motifs to genes
```{r}
network <- inner_join(closest_genes, motifPeaks, by = c("query_region" = "peak"))
network <- inner_join(network, motifs19.5[,c("motif", "motif.name")], by = c("motif" = "motif"))
network <- network[network$gene_name %in% markers,]
```

Grouping motifs of the same gene families together
```{r}
jaspar <- read.table("/work/LAS/geetu-lab/hhvu/JASPAR-library.txt", header = T, sep = "\t")
jaspar <- jaspar[jaspar$ID %in% network$motif,]
network <- inner_join(network, jaspar[,c("ID", "Family")], by = c("motif" = "ID"))
keep <- network$Family
keep <- subColon(keep)
for (i in 1:length(keep)) {
  keep[i] <- list(unique(keep[i][[1]]))
  for (j in setdiff(1:length(keep), which(keep %in% keep[i]))) {
    if (sum(keep[j][[1]] %in% keep[i][[1]]) > 0) {
      keep[j] <- keep[i]
    }
  }
}
keep2 <- unlist(lapply(keep, toString))
network$Family2 <- keep2

all <- data.frame(tfGene=character(), regGene=character())
for (i in unique(network$Family2)) {
  tfGene <- unique(network[which(network$Family2 == i), "motif.name"])
  tf <- toString(tfGene)
  regGene <- unique(network[which(network$Family2 == i), "gene_name"])
  tbl <- data.frame(matrix(nrow = length(tfGene)*length(regGene), ncol = 2))
  for (j in 1:length(tfGene)) {
    tbl[((j-1)*length(regGene)+1):(j*length(regGene)),1] <- tf
    tbl[((j-1)*length(regGene)+1):(j*length(regGene)),2] <- as.character(regGene)
  }
  all <- rbind(all, tbl)
}
all$X2 <- toupper(all$X2)
all <- dplyr::distinct(all)

g <- unique(all$X2)
for (k in 1:length(g)) {
  if (length(grep(g[k], all$X1)) != 0) {
    all[which(g[k] == all$X2), "X2"] <- all[grep(g[k], all$X1)[1], "X1"]
  }
}
all <- dplyr::distinct(all)

#write.table(all, "/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/noConns-commonPeaksGenes-gd9.5.txt", sep = "\t", row.names = F, quote = F)
```

Build network in Cytoscape.
```{r}
knitr::include_graphics("Files/noConns-PeaksGenes-gd19.5.PNG")
```

```{r}
sort(table(all$X2))[(length(table(all$X2))-15):length(table(all$X2))]
```

```{r}
cols <- c("#3e5d66", "#3e5d66")

cov_plot <- CoveragePlot(
  object = unintegrated,
  region = "16-83488834-83489632",
  feature = "Col4a1",
  annotation = TRUE,
  peaks = TRUE,
  heights = c(6, 1, 1),
  extend.upstream = 5000,
  extend.downstream = 5000,
  idents = c("GD15.5_Trophoblast_cells", "GD19.5_Trophoblast_cells")
)

cov_plot


cov_plot <- CoveragePlot(
  object = unintegrated,
  region = "1-12608710-12609256",
  feature = "Cited2",
  annotation = TRUE,
  peaks = TRUE,
  heights = c(6, 1, 1),
  extend.upstream = 5000,
  extend.downstream = 5000,
  idents = c("GD15.5_Trophoblast_cells", "GD19.5_Trophoblast_cells")
)

cov_plot


```



