---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 15,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>

```{r}
library("org.Rn.eg.db")
library("clusterProfiler")

go <- function(genes) {
  GO <- enrichGO(gene = genes, OrgDb=org.Rn.eg.db, ont = "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.05, maxGSSize=500, readable = T, keyType = "ENSEMBL")
  goSimplify <- simplify(GO)
  res <- as.data.frame(goSimplify)
  #res$Rank <- seq(1, nrow(res))
  res$GeneRatio <- sapply(strsplit(res$GeneRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$BgRatio <- sapply(strsplit(res$BgRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$Fold <- as.numeric(res$GeneRatio)/as.numeric(res$BgRatio)
  res <- res[res$qvalue <= 0.05 & res$Fold >= 2 & res$Count >= 5,]
  
  return(res)
}
```


## GD 15.5:
```{r}
library(Signac)
library(Seurat)
library(ggplot2)
library(ggpubr)
library(GenomicRanges)
library(patchwork)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5_res0.8.rda")

exp <- AverageExpression(data)
exp <- as.data.frame(exp$RNA)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
```


Invasive trophoblast cells: <br>
Genes with >= 3 peaks tend to be higher expressed:
```{r}
closest_genes <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc-toGenes.txt", sep = "\t")
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "23")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, invasive trophoblast GD15.5")

print("Number of genes with >= 3 peaks:")
length(genes)
print("Number of genes with < 3 peaks:")
length(unique(df[df$category == "< 3 peaks", "gene"]))
```

TBC non-specific peaks:
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]

rats1.marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-unfilteredMarkers.txt", header = T, sep = "\t", quote = "", stringsAsFactors = T, fill = T)
spePeaks <- rats1.marker[rats1.marker$cluster == "Invasive trophoblast cells" & rats1.marker$avg_log2FC >= log2(1.5) & rats1.marker$p_val_adj <= 0.05,]
nonspePeaks <- setdiff(rownames(atac3), spePeaks$gene)
length(nonspePeaks)

closest_genes <- ClosestFeature(rats, regions = nonspePeaks)

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "23")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, TBC-nonspecific peaks GD15.5")

print("Number of genes with >= 3 peaks:")
length(genes)
print("Number of genes with < 3 peaks:")
length(unique(df[df$category == "< 3 peaks", "gene"]))

goGenes3Peaks <- go(rnor6[rnor6$gene_name %in% genes, 1])
goGenes3Peaks[grep("epithe|migration|branching|placenta|embryo", goGenes3Peaks$Description),]

intersect(genes, marker$gene) #cited2 has 3 non-specific peaks
print("Number of markers with >= 3 non-specific peaks:")
length(intersect(genes, marker$gene))


goMarkers3Peaks <- go(rnor6[rnor6$gene_name %in% intersect(genes, marker$gene), 1])
goMarkers3Peaks[grep("epithe|migration|branching|placenta|embryo", goMarkers3Peaks$Description),]
```
So with peaks not TBC specific, the same trend is observed: genes with more peaks tend to higher expressed. <br>

TBC markers with TBC-nonspecific peaks: 326 out of 394 markers have non-specific peaks.
```{r}
markersWithNonspePeaks <- intersect(df$gene, marker$gene)
length(markersWithNonspePeaks)

markersWithNonspePeaks

markersWithNonspePeaks15.5 <- markersWithNonspePeaks
```
Expression of TBC markers with and without non-specific peaks:
```{r}
dat <- exp[exp$gene %in% df$gene, c("gene", "23")]
dat$cat <- ifelse(dat$gene %in% markersWithNonspePeaks, "markersWithNonspePeaks", "markersWithoutNonspePeaks")

wilcox.test(dat[dat$cat == "markersWithNonspePeaks", "23"], dat[dat$cat == "markersWithoutNonspePeaks", "23"])

ggplot(dat, aes(x=cat, y=log10(`23`+10^(-5)))) + geom_boxplot() + ggtitle("Expression of markers with and without TBC-nonspecific peaks GD15.5")
```

Gene ontology of markers with and without non-specific peaks:
```{r}
with <- go(rnor6[rnor6$gene_name %in% markersWithNonspePeaks, 1])
without <- go(rnor6[rnor6$gene_name %in% dat[dat$cat == "markersWithoutNonspePeaks", 1], 1])

with[grep("epithe|migration|branching|placenta|embryo", with$Description),]

without[grep("epithe|migration|branching|placenta|embryo", without$Description),] 
```


## GD 19.5:
```{r}
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-4-5-6-7_res0.8.rda")
data2

exp <- AverageExpression(data2)
exp <- as.data.frame(exp$RNA)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
```

Invasive trophoblast cells: <br>
```{r}
closest_genes <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc-toGenes.txt", sep = "\t")
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T)
```

Genes with >= 3 peaks tend to be higher expressed:
```{r}
t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "6")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)
df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, invasive trophoblast GD19.5")

print("Number of genes with >= 3 peaks:")
length(genes)
print("Number of genes with < 3 peaks:")
length(unique(df[df$category == "< 3 peaks", "gene"]))
```


TBC non-specific peaks:
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]

rats1.marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-unfilteredMarkers.txt", header = T, sep = "\t", quote = "", stringsAsFactors = T, fill = T)
spePeaks <- rats1.marker[rats1.marker$cluster == "Invasive trophoblast cells" & rats1.marker$avg_log2FC >= log2(1.5) & rats1.marker$p_val_adj <= 0.05,]
nonspePeaks <- setdiff(rownames(atac3), spePeaks$gene)
length(nonspePeaks)

closest_genes <- ClosestFeature(rats, regions = nonspePeaks)

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "6")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, TBC-nonspecific peaks GD19.5")

print("Number of genes with >= 3 peaks:")
length(genes)

goGenes3Peaks <- go(rnor6[rnor6$gene_name %in% genes, 1])
goGenes3Peaks[grep("epithe|migration|branching|placenta|embryo", goGenes3Peaks$Description),]

intersect(genes, marker$gene) #cited2 has 2 non-specific peaks
length(intersect(genes, marker$gene))

goMarkers3Peaks <- go(rnor6[rnor6$gene_name %in% intersect(genes, marker$gene), 1])
goMarkers3Peaks[grep("epithe|migration|branching|placenta|embryo", goMarkers3Peaks$Description),]
```

TBC markers with TBC-nonspecific peaks are TBC markers:
```{r}
markersWithNonspePeaks <- intersect(df$gene, marker$gene)
length(markersWithNonspePeaks)

markersWithNonspePeaks 
markersWithNonspePeaks19.5 <- markersWithNonspePeaks
```
Expression of TBC markers with and without non-specific peaks:
```{r}
dat <- exp[exp$gene %in% df$gene, c("gene", "6")]
dat$cat <- ifelse(dat$gene %in% markersWithNonspePeaks, "markersWithNonspePeaks", "markersWithoutNonspePeaks")

wilcox.test(dat[dat$cat == "markersWithNonspePeaks", "6"], dat[dat$cat == "markersWithoutNonspePeaks", "6"])

ggplot(dat, aes(x=cat, y=log10(`6`+10^(-5)))) + geom_boxplot() + ggtitle("Expression of markers with and without TBC-nonspecific peaks GD19.5")
```

Gene ontology of markers with and without non-specific peaks: genes with non-specific peaks (having peaks everywhere) seem to have general functions (placenta development, differentiation), but genes with only specific peaks seem to have specific functions of the cell type.
```{r}
with <- go(rnor6[rnor6$gene_name %in% markersWithNonspePeaks, 1])
without <- go(rnor6[rnor6$gene_name %in% dat[dat$cat == "markersWithoutNonspePeaks", 1], 1])

with[grep("epithe|migration|branching|placenta|embryo", with$Description),]

without[grep("epithe|migration|branching|placenta|embryo", without$Description),] 
```


# Misc
What and how many genes have non-specific peaks at both timepoints:
```{r}
length(intersect(markersWithNonspePeaks19.5, markersWithNonspePeaks15.5))

intersect(markersWithNonspePeaks19.5, markersWithNonspePeaks15.5)
```

```{r}
sessionInfo()
```
