---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 15,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>

```{r}
library("org.Rn.eg.db")
library("clusterProfiler")

go <- function(genes) {
  GO <- enrichGO(gene = genes, OrgDb=org.Rn.eg.db, ont = "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.05, maxGSSize=500, readable = T, keyType = "ENSEMBL")
  goSimplify <- simplify(GO)
  res <- as.data.frame(goSimplify)
  #res$Rank <- seq(1, nrow(res))
  res$GeneRatio <- sapply(strsplit(res$GeneRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$BgRatio <- sapply(strsplit(res$BgRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$Fold <- as.numeric(res$GeneRatio)/as.numeric(res$BgRatio)
  res <- res[res$qvalue <= 0.05 & res$Fold >= 2 & res$Count >= 5,]
  
  return(res)
}
```

# gd 15.5 ver 2

```{r}
library(Signac)
library(Seurat)
library(ggplot2)
library(ggpubr)
library(GenomicRanges)
library(patchwork)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5_res0.8.rda")

exp <- AverageExpression(data)
exp <- as.data.frame(exp$RNA)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
```

What if I use all TBC (specific or not) peaks together, genes are markers or not:
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

closest_genes <- ClosestFeature(rats, regions = rownames(atac3))

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "23")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, all peaks all genes GD15.5")

print("Number of genes with >= 3 peaks:")
length(genes)
print("Number of genes with < 3 peaks:")
length(unique(df[df$category == "< 3 peaks", "gene"]))
```
Within markers, the same is not observed (genes with >= 3 peaks or < 3 peaks are equally expressed). I think it's because markers are highly expressed already.
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

closest_genes <- ClosestFeature(rats, regions = rownames(atac3))

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "23")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)

df <- df[df$gene %in% marker$gene,]

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, all peaks marker genes GD15.5")
```

Peak distribution of marker genes:
```{r}
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)
t2 <- t[intersect(names(t),marker$gene)]
non <- setdiff(marker$gene, names(t))
nont <- rep(0, length(non))
names(nont) <- non
t3 <- as.vector(t2)
names(t3) <- names(t2)
t3 <- c(t3, nont)
hist(t3, breaks = seq(0, 20, 1))

length(t3[t3 >= 3])
```
202 out of 394 marker genes have >= 3 peaks. <br>
Of the markers with peaks, how many peaks are specific
```{r}
specificPeaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc-toGenes.txt", sep = "\t")
df2 <- closest_genes[closest_genes$gene_name %in% marker$gene,]

#number of specific peaks associated to markers
length(intersect(df2$query_region, specificPeaks$query_region))

#number of peaks associated to markers
length(df2$query_region)

#number of non-specific peaks
length(setdiff(closest_genes$query_region, specificPeaks$query_region))

dat <- data.frame(marker=c(423,1345-423), nonmarker=c(3842-423,18760-(3842-423)))
rownames(dat) <- c("specific", "nonspecific")
dat
fisher.test(dat, alternative = "greater")
```

Markers that only have nonspecific peaks:
```{r}
nonSpe <- df2[df2$query_region %in% setdiff(closest_genes$query_region, specificPeaks$query_region),]
onlyNonSpe <- nonSpe[!(nonSpe$gene_name %in% specificPeaks$gene_name),]

go_onlyNonSpe <- go(rnor6[rnor6$gene_name %in% onlyNonSpe$gene_name, 1])

go_onlyNonSpe[grep("epith|placenta|actin|vascul", go_onlyNonSpe$Description),]
```

Markers that only have specific peaks:
```{r}
onlySpe <- specificPeaks[!(specificPeaks$gene_name %in% nonSpe$gene_name),]
go_onlySpe <- go(rnor6[rnor6$gene_name %in% onlySpe$gene_name, 1])

go_onlySpe[grep("epith|placenta|actin|vascul", go_onlySpe$Description),]
```

Markers with both:
```{r}
both <- specificPeaks[specificPeaks$gene_name %in% intersect(specificPeaks$gene_name, nonSpe$gene_name),]
go_both <- go(rnor6[rnor6$gene_name %in% both$gene_name, 1])

go_both[grep("epith|placenta|actin|vascul", go_both$Description),]
```

## GD 19.5 ver 2:

```{r}
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-4-5-6-7_res0.8.rda")
data2

exp <- AverageExpression(data2)
exp <- as.data.frame(exp$RNA)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
```
What if I use all TBC (specific or not) peaks together, genes are markers or not:
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

closest_genes <- ClosestFeature(rats, regions = rownames(atac3))

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "6")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, all peaks all genes GD19.5")

print("Number of genes with >= 3 peaks:")
length(genes)
print("Number of genes with < 3 peaks:")
length(unique(df[df$category == "< 3 peaks", "gene"]))
```

Markers with >= 3 peaks are not higher expressed than markers with < 3 peaks. Maybe because they are already highly expressed.
```{r}
c <- rownames(rats@meta.data[rats@meta.data$predicted.id == "Invasive trophoblast cells",])
atac2 <- rats@assays$MACSpeaks[,c]
atac3 <- atac2[rowMeans(atac2) > 0,]
atac3 <- atac2[rowSums(atac2 > 0) >= ncol(atac2)/10, ]
dim(atac3)

closest_genes <- ClosestFeature(rats, regions = rownames(atac3))

t <- table(closest_genes$gene_name)

genes <- names(t[t >= 3])

exp <- exp[order(rownames(exp)),]
exp$gene <- rownames(exp)
df <- exp[exp$gene %in% names(t), c("gene", "6")]
colnames(df) <- c("gene", "value")
df2 <- data.frame(gene = setdiff(names(t), df$gene), value = rep(0, length(setdiff(names(t), df$gene))))
names(df2) <- colnames(df)
df <- rbind(df, df2)

df$category <- ifelse(df$gene %in% genes, ">= 3 peaks", "< 3 peaks")

marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T)

df <- df[df$gene %in% marker$gene,]

wilcox.test(df[df$category == ">= 3 peaks", "value"], df[df$category == "< 3 peaks", "value"])

ggplot(df, aes(x=category, y=log10(value+10^(-5)))) + geom_boxplot() + ggtitle("Cutoff = 3, all peaks marker genes GD19.5")

```

Peak distribution of marker genes:
```{r}
marker <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T)
t2 <- t[intersect(names(t),marker$gene)]
non <- setdiff(marker$gene, names(t))
nont <- rep(0, length(non))
names(nont) <- non
t3 <- as.vector(t2)
names(t3) <- names(t2)
t3 <- c(t3, nont)
hist(t3, breaks = seq(0, 12, 1))

length(t3[t3 >= 3])
```
Of the markers with peaks, how many peaks are specific
```{r}
specificPeaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc-toGenes.txt", sep = "\t")
df2 <- closest_genes[closest_genes$gene_name %in% marker$gene,]

#number of specific peaks
length(specificPeaks$query_region)

#number of specific peaks associated to markers
length(intersect(df2$query_region, specificPeaks$query_region))

#number of peaks associated to markers
length(df2$query_region)

#number of non-specific peaks
length(setdiff(closest_genes$query_region, specificPeaks$query_region))

dat <- data.frame(marker=c(339,765-339), nonmarker=c(2033-339,7808-(2033-339)))
rownames(dat) <- c("specific", "nonspecific")
dat
fisher.test(dat, alternative = "greater")
```

Markers that only have nonspecific peaks:
```{r}
nonSpe <- df2[df2$query_region %in% setdiff(closest_genes$query_region, specificPeaks$query_region),]
onlyNonSpe <- nonSpe[!(nonSpe$gene_name %in% specificPeaks$gene_name),]

go_onlyNonSpe <- go(rnor6[rnor6$gene_name %in% onlyNonSpe$gene_name, 1])

go_onlyNonSpe[grep("epith|placenta|actin|vascul", go_onlyNonSpe$Description),]

go_onlyNonSpe$Description
```

Markers that only have specific peaks:
```{r}
onlySpe <- specificPeaks[!(specificPeaks$gene_name %in% nonSpe$gene_name),]
go_onlySpe <- go(rnor6[rnor6$gene_name %in% onlySpe$gene_name, 1])

go_onlySpe[grep("epith|placenta|actin|vascul", go_onlySpe$Description),]
```

Markers with both:
```{r}
both <- specificPeaks[specificPeaks$gene_name %in% intersect(specificPeaks$gene_name, nonSpe$gene_name),]
go_both <- go(rnor6[rnor6$gene_name %in% both$gene_name, 1])

go_both[grep("epith|placenta|actin|vascul", go_both$Description),]
```


```{r}
sessionInfo()
```
