---
title: "scATAC-seq Rat Metrial Glands"
author: "Ha T. H. Vu"
output: html_document
---
  
```{r setup, include=FALSE}
options(max.print = "75")
knitr::opts_chunk$set(
  echo = TRUE,
  collapse = TRUE,
  comment = "#>",
  fig.path = "Files/",
  fig.width = 15,
  prompt = FALSE,
  tidy = FALSE,
  message = FALSE,
  warning = TRUE
)
knitr::opts_knit$set(width = 75)
```

This is a documentation for analyses of scATAC-seq data, generated from rat metrial gland tissues on gestational day (GD) 15.5 and 19.5. <br>
  
```{r}
library("org.Rn.eg.db")
library("clusterProfiler")

go <- function(genes) {
  GO <- enrichGO(gene = genes, OrgDb=org.Rn.eg.db, ont = "BP", pvalueCutoff = 0.05, qvalueCutoff = 0.05, maxGSSize=500, readable = T, keyType = "ENSEMBL")
  goSimplify <- simplify(GO)
  res <- as.data.frame(goSimplify)
  #res$Rank <- seq(1, nrow(res))
  res$GeneRatio <- sapply(strsplit(res$GeneRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$BgRatio <- sapply(strsplit(res$BgRatio, "/"), function(x) as.numeric(x[1])/as.numeric(x[2]))
  res$Fold <- as.numeric(res$GeneRatio)/as.numeric(res$BgRatio)
  res <- res[res$qvalue <= 0.05 & res$Fold >= 2 & res$Count >= 5,]
  
  return(res)
}
```


# All TBC peaks, gd15.5 and gd19.5
GD15.5 with all genes:
```{r}
gd15.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc-toGenes.txt", header = T)
g <- go(gd15.5$ensembl_gene_id)
g$Description[1:15]
```
GD15.5 with gene markers of TBC:
```{r}
gd15.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-tbc-toGenes.txt", header = T)
tbc <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd15.5-wilcoxin-cluster23.txt", header = T)
g <- go(gd15.5[gd15.5$gene_name %in% tbc$gene, "ensembl_gene_id"])
g$Description[1:15]
```
GD19.5 with all genes:
```{r}
gd19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc-toGenes.txt", header = T)
g <- go(gd19.5$ensembl_gene_id)
g$Description[1:15]
```
GD19.5 with gene markers of TBC:
```{r}
gd19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-tbc-toGenes.txt", header = T)
tbc <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/scRNA-seqFiles/gd19.5-wilcoxin-cluster6.txt", header = T)
g <- go(gd19.5[gd19.5$gene_name %in% tbc$gene, "ensembl_gene_id"])
g$Description[1:15]
```

# TBC peaks common between gd15.5 and gd19.5
Using GD19.5 coordinates, associate the common peaks to the nearest genes, then run GO analysis.
```{r}
library(Signac)
library(Seurat)
library(JASPAR2020)
library(TFBSTools)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
library(patchwork)
library(dplyr)
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
peaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks-gd19.5.bed", header=F)
closest_genes <- ClosestFeature(rats, regions = peaks$V4)

g <- go(closest_genes$ensembl_gene_id)
g$Description[1:15]
```

# DA peaks at both timepoints:
```{r}
new15.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/daPeaks/gd15.5TBC.txt", header = T)
gd15.5GO <- go(unique(new15.5$ensembl_gene_id))
dim(gd15.5GO)

new19.5 <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/daPeaks/gd19.5TBC.txt", header = T)
gd19.5GO <- go(unique(new19.5$ensembl_gene_id))
gd19.5GO$Description
```

# Conserved peaks
```{r}
library(Signac)
library(Seurat)
library(JASPAR2020)
library(TFBSTools)
library(BSgenome.Rnorvegicus.Ensembl.rn6)
library(patchwork)
library(dplyr)
set.seed(1234)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/rnor6.ranges.rda")
load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd19.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
peaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/commonPeaks-gd19.5-hg38Coor-inEVT.bed", header=F)
closest_genes <- ClosestFeature(rats, regions = peaks$V4)

da19.5peaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/daPeaks/da19.5-hg38Coor-inEVT.bed", header = F)
closest_genes2 <- ClosestFeature(rats, regions = da19.5peaks$V4)

load("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/4_RNAintegration/2_MACS2Peaks/annotation_scRNAseq/gd15.5-ATACwithRNAlables.rda")
Idents(rats) <- rats@meta.data$predicted.id # change from old identities to those predicted by scRNA-seq
DefaultAssay(rats) <- 'MACSpeaks'
Annotation(rats) <- rnor6.ranges
da15.5peaks <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/daPeaks/da15.5-hg38Coor-inEVT.bed", header = F)
closest_genes3 <- ClosestFeature(rats, regions = da15.5peaks$V4)

genes <- unique(c(closest_genes$ensembl_gene_id, closest_genes2$ensembl_gene_id, closest_genes3$ensembl_gene_id))
g <- go(genes)
g$Description[1:15]

conservedRats <- read.table("/work/LAS/geetu-lab/hhvu/project3_scATAC/scATAC-seq-analysis/5_integrating2timepoints/annotation_scRNAseq/commonPeaks/conservedRat.txt", header = T, sep = "\t")
genes2 <- unique(c(
  closest_genes[toupper(closest_genes$gene_name) %in% conservedRats$ratGenes, "ensembl_gene_id"], 
  closest_genes2[toupper(closest_genes2$gene_name) %in% conservedRats$ratGenes, "ensembl_gene_id"], 
  closest_genes3[toupper(closest_genes3$gene_name) %in% conservedRats$ratGenes, "ensembl_gene_id"]))
genes2.2 <- unique(c(
  closest_genes[toupper(closest_genes$gene_name) %in% conservedRats$ratGenes, "gene_name"], 
  closest_genes2[toupper(closest_genes2$gene_name) %in% conservedRats$ratGenes, "gene_name"], 
  closest_genes3[toupper(closest_genes3$gene_name) %in% conservedRats$ratGenes, "gene_name"]))
g2 <- go(genes2)

```